File: Server/getTabBuddies.php
<?php
require "globalFunctions.php";
$pdo = new PDO("pgsql:host=" . HOST . ";port=" . PORT . ";dbname=" . DATABASE . ";user=".USERNAME . ";password=" . PASSWORD);

$myId = $_POST["myId"];
$token = $_POST["token"];
$gender = (int) $_POST["gender"]; // -1 = all, 0 = male, 1 = female, 2 = other
$minAge = (int) $_POST["minAge"];
$maxAge = (int) $_POST["maxAge"];
$country = (int) $_POST["country"]; // -1 = all, 0 = Afghanistan, ... 196 = Zimbabwe
$sort = (int) $_POST["sort"]; // 0 = active, 1 = new
$interests = $_POST["interests"]; // interests example: "&C++&&Python&&SwiftUI&"
$bottomLastVisitedAt = $_POST["bottomLastVisitedAt"];
$bottomSignedUpAt = $_POST["bottomSignedUpAt"];

if (empty($myId)) {
    $myId = $emptyUuid;
}

$query = "insert into time_format_2 (user_id, bottom_last_visited_at_string) values (?, ?);";
$stmt = $pdo->prepare($query);
$stmt->execute(array($myId, $bottomLastVisitedAt));

$isValid =
    ($myId === $emptyUuid || isAuthenticated($myId, $token)) &&
    isValidInt($gender, -1, 2) &&
    isValidInt($minAge, 13, 130) &&
    isValidInt($maxAge, $minAge, 130) &&
    isValidInt($country, -1, 196) &&
    isValidInterests($interests) &&
    isValidInt($sort, 0, 1) &&
    isValidTime($bottomLastVisitedAt) &&
    isValidTime($bottomSignedUpAt);
if (!$isValid) {
    $pdo = null;
    die("Invalid");
}

$query = "select user_id, username, birthday, gender, country, intro, last_visited_at, signed_up_at from account where not user_id = ? and banned_at is null and disabled_at is null and deleted_at is null ";
$paramsArray = array($myId);

if ($gender !== -1) {
    $query .= "and gender = ? ";
    array_push($paramsArray, $gender);
}

if ($minAge !== 13 || $maxAge !== 130) {
    $maxAge += 1;
    $minDate = date("Y-m-d", strtotime("-{$maxAge} years"));
    $maxDate = date("Y-m-d", strtotime("-{$minAge} years"));
    $query .= "and birthday between ? and ? ";
    array_push($paramsArray, $minDate, $maxDate);
}

if ($country !== -1) {
    $query .= "and country = ? ";
    array_push($paramsArray, $country);
}

// Select buddies with OR interests.
if (strlen($interests)) {
    $query .= "and (false ";

    $interests = substr($interests, 1, -1);
    $interestsArray = explode("&&", $interests);
    foreach ($interestsArray as $interest) {
        $query .= "or interests like ? ";
        array_push($paramsArray, "%&{$interest}&%");
    }

    $query .= ") ";
}

// Select buddies with AND interests.
/*if (strlen($interests)) {
$query .= "and interests like ? ";
$likeQuery = "%";

$interests = substr($interests, 1, -1);
$interestsArray = explode("&&", $interests);
foreach ($interestsArray as $interest) {
$likeQuery .= "&{$interest}&%";
}
array_push($paramsArray, $likeQuery);
}*/

if ($sort === 0) {
    # Hide new users who made their accounts up to 1 day ago
    # in order to prevent new users from cluttering active users sort and prevent mass spam of new user account.
    //$query .= "and last_visited_at < ? and signed_up_at < current_date - interval '1 days' order by last_visited_at desc limit 20;";
    $query .= "and last_visited_at < ? order by last_visited_at desc limit 20;";
    array_push($paramsArray, $bottomLastVisitedAt);
} else {
    $query .= "and signed_up_at < ? order by signed_up_at desc limit 20;";
    array_push($paramsArray, $bottomSignedUpAt);
}

$stmt = $pdo->prepare($query);
$stmt->execute($paramsArray);

$counter = 1; // PHP arrays start from 1
$rows = array();
while ($row = $stmt->fetch()) {
    $rows[$counter++] = array(
        "buddyId" => $row[0],
        "username" => $row[1],
        "birthday" => $row[2],
        "gender" => $row[3],
        "country" => $row[4],
        "intro" => $row[5],
        "lastVisitedAt" => $row[6],
    );

    $bottomLastVisitedAt = $row[6];
    $bottomSignedUpAt = $row[7];
}

$rows[$counter++] = array(
    "bottomLastVisitedAt" => $bottomLastVisitedAt,
    "bottomSignedUpAt" => $bottomSignedUpAt,
);

echo json_encode($rows);

$pdo = null;

/*echo "QUERY: {$query}\n";
echo "ARRAY: ";
print_r($paramsArray);*/

File: Server/getTabBuddies.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"
)

func getTabBuddies(w http.ResponseWriter, r *http.Request) {
	r.ParseForm()
	token := r.Form.Get("token")
	gender := r.Form.Get("gender")
	minAge := r.Form.Get("minAge")
	maxAge := r.Form.Get("maxAge")
	country := r.Form.Get("country")
	sort := r.Form.Get("sort")
	interests := r.Form.Get("interests")
	bottomLastVisitedAt := r.Form.Get("bottomLastVisitedAt")
	bottomSignedUpAt := r.Form.Get("bottomSignedUpAt")
	myId := getMyId(token)

	isValid :=
		(myId == uuidNil || isAuthenticated(myId)) &&
			isValidInt(gender, -1, 2) &&
			isValidInt(minAge, 13, 130) &&
			isValidInt(maxAge, toInt(minAge), 130) &&
			isValidInt(country, -1, 196) &&
			isValidInterests(interests) &&
			isValidInt(sort, 0, 1) &&
			isValidTime(bottomLastVisitedAt) &&
			isValidTime(bottomSignedUpAt)
	if !isValid {
		fmt.Fprintln(w, "Invalid")
		return
	}

	query := `
	select user_id,
           username,
           gender,
           birthday,
           country,
           intro,
           last_visited_at,
           signed_up_at
    from   account
    where  not user_id = $1
           and banned_at is null
           and disabled_at is null
           and deleted_at is null `

	var params []interface{}
	params = append(params, myId)

	if gender != "-1" {
		query += fmt.Sprintf("and gender = $%d ", len(params)+1)
		params = append(params, gender)
	}

	if minAge != "13" || maxAge != "130" {
		minDate := time.Now().AddDate(-toInt(maxAge)-1, 0, 0)
		maxDate := time.Now().AddDate(-toInt(minAge), 0, 0)

		query += fmt.Sprintf("and birthday between $%d and $%d ", len(params)+1, len(params)+2)
		params = append(params, minDate, maxDate)
	}

	if country != "-1" {
		query += fmt.Sprintf("and country = $%d ", len(params)+1)
		params = append(params, country)
	}

	if interests != "" {
		query += "and (false "

		noFirstOrLast := interests[1 : len(interests)-1]
		interests := strings.Split(noFirstOrLast, "&&")

		for _, interest := range interests {
			query += fmt.Sprintf("or interests like '%%' || '&' || $%d || '&' || '%%' ", len(params)+1)
			params = append(params, interest)
		}

		query += ") "
	}

	if sort == "0" {
		// Hide new users who made their accounts up to 1 day ago
		// in order to prevent new users from cluttering active users sort and prevent mass spam of new user account.
		//query += "and last_visited_at < ? and signed_up_at < current_date - interval '1 days' order by last_visited_at desc limit 20;";
		query += fmt.Sprintf("and last_visited_at < $%d order by last_visited_at desc limit 20;", len(params)+1)
		params = append(params, bottomLastVisitedAt)
	} else {
		query += fmt.Sprintf("and signed_up_at < $%d order by signed_up_at desc limit 20;", len(params)+1)
		params = append(params, bottomSignedUpAt)
	}

	rows, err := pool.Query(context.Background(), query, params...)
	handleError(err)
	defer rows.Close()

	mapper := map[int]interface{}{}
	newBottomLastVisitedAt := toTime(bottomLastVisitedAt)
	newBottomSignedUpAt := toTime(bottomSignedUpAt)

	for rows.Next() {
		var buddyId string
		var username string
		var gender int
		var birthday time.Time
		var country int
		var intro string
		var lastVisitedAt time.Time
		var signedUpAt time.Time

		err = rows.Scan(&buddyId, &username, &gender, &birthday, &country, &intro, &lastVisitedAt, &signedUpAt)
		handleError(err)

		newBottomLastVisitedAt = lastVisitedAt
		newBottomSignedUpAt = signedUpAt

		newMapper := map[string]interface{}{
			"buddyId":       buddyId,
			"username":      username,
			"gender":        gender,
			"birthday":      birthday,
			"country":       country,
			"intro":         intro,
			"lastVisitedAt": lastVisitedAt,
		}
		mapper[len(mapper)] = newMapper
	}
	handleError(rows.Err())

	newMapper := map[string]interface{}{
		"bottomLastVisitedAt": newBottomLastVisitedAt,
		"bottomSignedUpAt":    newBottomSignedUpAt,
	}
	mapper[len(mapper)] = newMapper

	json, err := json.Marshal(mapper)
	handleError(err)

	w.Header().Set("Content-Type", "application/json")
	w.Write(json)
}
